cli: simplify and improve dump performance

madelynnblue opened the pull request:
A significant percentage (near 50%) of CPU time during dump is spent
converting strings returned from the SELECT into SQL-safe strings and
writing that to the output. Much of this work can be done concurrently,
which improves the performance, and gets it close to that of piping
a `SELECT *` to a file. This is done using an errgroup and passing
various stages over channels.

We can also greatly simplify dump now since cockroach streams
results. There's no need anymore to page through the data.

See cockroachdb/docs#1674

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18472)
<!-- Reviewable:end -->

knz reviewed (commented):
I think this is fine, but don't we want to throttle this? I don't recall that Go guarantees unlimited concurrent goroutines. @petermattis what do you think?

mjibson commented:
Throttle what? There's only 3 go routines. It never went above 200% CPU during my tests (on a 12-core machine). The buffered chan only has ~100 strings in it. Am I missing something?

knz commented:
Ugh, I'm sorry for the noise. I had wrongly read that there was one goroutine per row. LGTM as-is.

madelynnblue committed:
cli: simplify and improve dump performance

A significant percentage (near 50%) of CPU time during dump is spent
converting strings returned from the SELECT into SQL-safe strings and
writing that to the output. Much of this work can be done concurrently,
which improves the performance, and gets it close to that of piping
a `SELECT *` to a file. This is done using an errgroup and passing
various stages over channels.

We can also greatly simplify dump now since cockroach streams
results. There's no need anymore to page through the data.

See cockroachdb/docs#1674
